Katello/katello
# Fixes #34501 - Refactor CV import to use MetadataMap + support cp_id product lookup

#### What are the changes introduced in this pull request?

The labels of Red Hat products can sometimes change. Given that the import process is driven by labels (looking up products in the importing org) this would break when the product labels of the exporting and importing org no longer match. While the solution would be to refresh the manifest in both orgs and re-export, this is undesirable. This change allows Red Hat products to be looked up by their cp_id when importing, if present. For backward compatibility, the product lookup will use the label from the export metadata if the cp_id is missing.

#### Considerations taken when implementing this change?

While looking into how to handle cp_id on the importing side I came across this method: 
https://github.com/Katello/katello/blob/master/app/services/katello/pulp3/content_view_version/import.rb#L64-L77

Given an export metadata hash, it produces two differently shaped hashes depending on the parameters given. Looking at this method and things like complexity introduced in areas where `metadata_map` was used, and the vague shape of the data (it's hard to tell what's inside) I created the MetadataMap class. Its role is to parse the export metadata json into a collection of objects that are easy to understand and pass around. It simplified a lot of the dynflow actions, import services code, and tests. More importantly it makes the code easier to follow and change in the future - especially if fields get added or removed from the metadata, or if its general shape should change.

I also cleaned up (what I found to be) encapsulation issues. There were a number of attr_accessors which should be attr_readers (or not present at all) and the ContentViewVersion::Import service class had a lot of methods moved from class methods to instance methods given that the org and metadata seemed to be very common arguments to pass to these, and the instance of Import was already aware of those values.

#### What are the testing steps for this pull request?

- Create an org for import and export
- Import a manifest in both orgs
- In the export org, enable a Red Hat repo (satellite tools is a good, small one)
- Export the repository

- Simulate a change of the label thru the rails console:
```
prod = Katello::Product.where(label: 'Red_Hat_Enterprise_Linux_for_x86_64', organization_id: $IMPORT_ORG_ID).first
prod.update_attribute(:label, prod.label + "-updated")
```
- Import the exported repository and notice the failure reported in the BZ: "The organization manifest does not contain the subscriptions required to enable the following repositories"

- Check out this PR and try the import again - it should succeed!

- Due to the extensive refactoring on the import side, I recommend validating the other workflows by importing various kinds of exports. I've already done so, but it would be good to double-check this.
